### ISSUE TITLE
Request: Support for automated patching

### ISSUE THREAD

**Issue by DeftNerd** [keylase/nvidia-patch#162]
My ideal workflow with this package would be an installer or bash script that clones the repo somewhere like /opt/nvidia-patch. Then it would symlink some systemd scripts and place a config file in /etc/default. 

The systemd script would run after the network is available. It would look at the currently loaded driver to see if it's already patched (perhaps the last patched version could be specified in the /etc/default/nvidia-patch config file?). If it's not patched, it would manually trigger another systemd script that normally does automatic daily updates of the /opt/nvidia-patch folder from the git repo. 

After the update is complete, it would patch the current driver (if possible)

This would allow people to have less anxiety that their well-crafted solution would be upset by driver updates. 

Do you think this is something that fits with this project, or do you prefer the goal of manual updates and execution of the program as needed?

**Comment by DeftNerd** [comment]
Oh, another possible useful trick is [using apt/dpkb hooks to execute scripts](https://www.cyberciti.biz/faq/debian-ubuntu-linux-hook-a-script-command-to-apt-get-upgrade-command/)

This could allow you to put a few scripts in there. 00nvidia-patch could pull a file from your repo through raw.githubcontent.com to see what versions of drivers have patches available, and specifically block versions of the driver that don't have patches available so they're held back on updates. 

You could also use it to call Apt::Update::Post-Invoke-Success to run a script to actually patch a compatible driver after its installed.

**Comment by Snawoot** [comment]
Hello!

It's a good idea! I'm aware about some deployments which already pull this patch via network. Good example is [Cloudbox](https://cloudbox.works/) (here is corresponding [Ansible recipe in it's code](https://github.com/Cloudbox/Cloudbox/blob/92c30f02d75e9a087d0bcc4ee517f756d3ea1f06/roles/nvidia/tasks/nvidia_patch.yml)). Also some users pull patch script in user script on Unraid.

There may be many approaches to follow updates of drivers and this patch. Some of them may have downsides which are not acceptable for some users. For example, your approach with systemd units may be considered not suitable in some cases because it relies on network resource, pulls untrusted code or just because it uses systemd (there are still many systems without it). Believe or not, we still have users with 7 years old Ubuntu 12.04: #30.

Approach with APT obviously will work only with Debian-derived systems which have Nvidia drivers installed by deb package. It's not always the case because many people prefer installing CUDA toolkit and drivers by a .run file. In this case they do not receive auto-updates for driver but also they don't care about driver version as far as it works. Back then, when I used NVENC it was my preferred option.

Good auto-update solution is tricky. Summarizing, we have to keep following items in mind:
* There are wide variety of Linux distros, including ancient ones.
* Driver may be installed by package manager or by standalone self-extracting package, so maintenance technique may differ.
* Driver may be not modifiable in system. For example, there is [nvidia-docker case](https://github.com/keylase/nvidia-patch#docker-support), where driver files mounted into container in read-only mode by special docker runtime. In such cases we messing up with dynamic linker to create override for driver without breaking container isolation.
* Network location of this patch may change if we'll experience some issues with Github.
* Driver update may be not a concern for user at all. Driver may be installed by standalone package once for entire operation term or host may not have direct access to Internet (a dedicated video processing farm case). For this reason auto-update setup has to be optional and not be included into basic patching workflow in order to keep things simple.

But it'll be nice to have such solution as an option. I'm thinking about some generic shell script, which relies only on basic system utilities. Also we can take steps to meet both solutions you outlined and make own auto-update implementation trivial for users. So, let's get to details.

>  It would look at the currently loaded driver to see if it's already patched (perhaps the last patched version could be specified in the /etc/default/nvidia-patch config file?). If it's not patched, it would manually trigger another systemd script that normally does automatic daily updates of the /opt/nvidia-patch folder from the git repo.

Usually applications store their state in somewhere in `/var/lib`. `/etc` is for configuration files edited by system administrator. But I think we can avoid relying on some stored state at all. If we make `patch.sh` invocation idempotent (i. e., subsequent of patch does nothing), we can just run latest patch.sh to ensure system is patched. I think best way to implement it is to consider current version patched when backup file exists and it's hashsum is equal to hashsum of target file. In all other cases we can proceed to usual patching logic. That will be very useful change which will be helpful in many cases. By the way, sha1 sum calculation already used in patch.sh, it'll be pretty easy to implement check.

> 00nvidia-patch could pull a file from your repo through raw.githubcontent.com to see what versions of drivers have patches available, and specifically block versions of the driver that don't have patches available so they're held back on updates.

I guess we need to implement additional patch.sh operation modes for this. patch.sh invoked with some command line key should respond with list of supported versions or return true/false exit code if it supports specific version passed via command line. This way we can check supported versions beforehand without shell file parsing or maintaining external config with version list.

Would you like to implement any of these features two or leave implementation to the others?

**Comment by Snawoot** [comment]
Ok, starting implementation of new operation modes and idempotent invocation.

**Comment by Snawoot** [comment]
I've just merged #163 which adds some interfaces to patch.sh in order to make it automation friendly. Now it is possible to query patch whether it supports specific driver version, get list of all supported versions and rely on patch itself to figure out whether patching required or library was already patched.

Now simpliest auto-update can be achieved with startup script like this:

```sh
tmp=$(mktemp)
wget -O "$tmp" https://raw.githubusercontent.com/keylase/nvidia-patch/master/patch.sh && \
bash "$tmp"
rm "$tmp"
```

More complicated scenarios may involve repo checkout/caching and other things which belong to matters of taste. Thus, at this point I consider we have all necessary for automated patching, just leaving nuanced details to end user.

**Comment by zimme** [comment]
@DeftNerd Did you ever end up putting together any apt/dpkg hook scripts for this?
